server/src-lib/Data/Parser/Json.hs

```diff
@@ -0,0 +1,221 @@
+  , toList
+  , fromList
+  , insert
+  , delete
+  , empty
+  , eitherDecode
+  , toEncJSON
+  , Data.Parser.Json.lookup
+  ) where
+
+import           Control.Applicative hiding (empty)
+import qualified Data.Aeson as J
+import           Data.Aeson.Parser (jstring)
+import           Data.Attoparsec.ByteString (Parser)
+import qualified Data.Attoparsec.ByteString as A
+import qualified Data.Attoparsec.ByteString.Char8 as A8
+import           Data.Bifunctor
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Lazy as L
+import           Data.Data
+import           Data.Functor
+import           Data.HashMap.Strict.InsOrd (InsOrdHashMap)
+import qualified Data.HashMap.Strict.InsOrd as OMap
+import           Data.Scientific
+import           Data.Text (Text)
+import qualified Data.Text as T
+import           Data.Vector (Vector)
+import qualified Data.Vector as V
+import           GHC.Generics
+import           Hasura.EncJSON
+import           Prelude
+import           Prelude hiding (error, undefined)
+
+--------------------------------------------------------------------------------
+-- Encoding via Hasura's EncJSON
+
+toEncJSON :: Value -> EncJSON
+toEncJSON =
+  \case
+    Object (Object_ omap) ->
+      encJFromAssocList (map (second toEncJSON) (OMap.toList omap))
+    Array vec -> encJFromList (map toEncJSON (V.toList vec))
+    String s -> encJFromJValue s
+    Number sci -> encJFromJValue sci
+    Bool b -> encJFromJValue b
+    Null -> encJFromJValue J.Null
+
+--------------------------------------------------------------------------------
+-- Copied constants from aeson
+
+#define BACKSLASH 92
+#define CLOSE_CURLY 125
+#define CLOSE_SQUARE 93
+#define COMMA 44
+#define DOUBLE_QUOTE 34
+#define OPEN_CURLY 123
+#define OPEN_SQUARE 91
+#define C_0 48
+#define C_9 57
+#define C_A 65
+#define C_F 70
+#define C_a 97
+#define C_f 102
+#define C_n 110
+#define C_t 116
+
+--------------------------------------------------------------------------------
+-- Our altered type
+
+-- | A JSON \"object\" (key\/value map). This is where this type
+-- differs to the 'aeson' package.
+newtype Object = Object_ { unObject_ :: InsOrdHashMap Text Value}
+  deriving (Eq, Read, Show, Typeable, Data, Generic)
+
+-- | Union the keys, ordered, in two maps, erroring on duplicates.
+union :: Object -> Object -> Either String Object
+union (Object_ x) (Object_ y) =
+  fmap
+    Object_
+    (traverse
+       id
+       (OMap.unionWithKey
+          (\k _a _b -> Left ("Duplicate key: " ++ T.unpack k))
+          (fmap Right x)
+          (fmap Right y)))
+
+-- | Empty object.
+empty :: Object
+empty = Object_ mempty
+
+-- | Insert before the element at index i. Think of it in terms of
+-- 'splitAt', which is (take k, drop k). Deletes existing key, if any.
+insert :: (Int, Text) -> Value -> Object -> Object
+insert (idx, key) val =
+  Object_ .
+  OMap.fromList .
+  uncurry (<>) .
+  second ((key, val) :) .
+  splitAt idx .
+  OMap.toList .
```

Comment:
Is it possible that we might lose the ordering of the keys here because of `OMap.toList`?

For e.g. Suppose these are the following operations:
```
do
  obj1 <- insert (0, key0) val0 obj0 
  obj2 <- insert (3, key3) val3 obj1
  obj3 <- insert (2, key2) val2 obj2
  obj4 <- insert (1, key1) val1 obj3
```
```
$ OMap.toList obj4 
[(key0, val0), (key1, val1), (key3, val3), (key2, val2)] 
```